File: summer2019/week4/next-greater-element-i.java
class Solution {
    /* Take an element in the first array and then 
     * find the same element in the second array 
     * and then look to the right in the second array 
     * to find a greater one.
     */
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> stack = new Stack();
        int[] result = new int[nums1.length];
        Arrays.fill(result, -1);
        
        // Populate hashmap with items in map and their indices
        Map<Integer, Integer> map = new HashMap();
        for (int i = 0; i < nums1.length; i++) {
            map.put(nums1[i], i);
        }
        
        // Iterate through nums2
        for (int i = 0; i < nums2.length; i++) {
            // If the most recent item added to the stack is less than the current item
            // check whether nums1 maps to the previous item because the current item
            // is its next greater element
            while(!stack.isEmpty() && stack.peek() < nums2[i]){
                Integer num = map.get(stack.pop());
                if(num != null){
                    result[num] = nums2[i];
                }
            }
            // Add each item in nums2 to stack, in order
            stack.push(nums2[i]);
        }
        return result;
    }
}

File: summer2021/unit4/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        result = []
        for num in nums1:
            found = False
            stack = nums2[::-1]
            while len(stack) >= 0:
                if len(stack) == 0:
                    result.append(-1)
                    break
                if not found:
                    if stack[-1] == num:
                        found = True
                    stack.pop()
                else:
                    if stack[-1] > num:
                        result.append(stack[-1])
                        break
                    else:
                        stack.pop()
        return result